### PR TITLE
Add cuda operators for Complex64 and Complex128 Tensor

### PR DESCRIPTION
**Pull request by MarioLulab**
### Original requirements
Some AI for Science related models(FNO, AFNO, PINO) use complex data types and related operations, So we need to support existing operations with Complex Tensor. 
We have previously supported some CPU operators for Complex Tensor, but still do not support more CUDA operators using Complex Tensor. This PR aims for supporting these.

### Main works
CPU ops for complex tensor related PR: https://github.com/Oneflow-Inc/oneflow/pull/10049
related issue: https://github.com/Oneflow-Inc/OneTeam/issues/1964

- broadcast_elementwise_binary
  | Op   |      complex type      |  Backend  |  Progress  |
  |:-----:|:-------------:|:------:|:------:|
  | Add |  cp64, cp128 | CUDA | Done |
  | Mul |  cp64, cp128   |  CUDA | Done |
  | Sub |  cp64, cp128 | CUDA | Done |
  | Equal |  cp64, cp128   |  CUDA | Done |
  | NotEqual |  cp64, cp128 | CUDA | Done |

  Operations that should not be supported by Complex numbers:
  Min, Max, Fmod, FloorDiv, TruncDiv, LessThan,
  bitwise operations, logical operations ...

- broadcast_elementwise_unary
  | Op   |      complex type      |  Backend  |  Progress   |
  |----------|:-------------:|:------:|:------:|
  | Cast |  cp64, cp128 | CUDA | Done |
  | Identity |  cp64, cp128 | CUDA | Done |

  Currently supports converting real tensor to complex tensor, and converting complex tensor to each other

- other exisiting operations
  | Op   |      complex type      |  Backend  |  Progress   |
  |----------|:-------------:|:------:|:------:|
  | slice |  cp64, cp128 | CUDA | Done |
  | constant_pad |    cp64, cp128   |  CUDA | Done |
  | reduce_sum |    cp64, cp128   |  CUDA | Done |

### future works
- broadcast_elementwise_binary
  Add support for `div`, `pow`

### CODE REVIEW

File: oneflow/core/common/data_type.cpp [hunk 1]
```diff
@@ -89,7 +89,7 @@ bool IsSupportRequireGradDataType(DataType data_type) {
   case type_proto: return true;
     OF_PP_FOR_EACH_TUPLE(
         REQUIRE_GRAD_CASE,
-        FLOATING_DATA_TYPE_SEQ FLOAT16_DATA_TYPE_SEQ BFLOAT16_DATA_TYPE_SEQ COMPLEX_DATA_TYPE_SEQ)
+        FLOATING_DATA_TYPE_SEQ FLOAT16_DATA_TYPE_SEQ BFLOAT16_DATA_TYPE_SEQ COMPLEX_DATA_TYPE_SEQ)    // TO-DO: if add cuComplex, will cause DataType::kComplex64 duplicate
```
Comment (levi131) [inline oneflow/core/common/data_type.cpp:92]: 这里检查的是`kComplex64 `，不需要再改COMPLEX_DATA_TYPE_SEQ

**Review by levi131 — COMMENTED** [review]
再加一些测试，可以加在test_complex.py中，和原先CPU部分测试放在一起

File: oneflow/core/ep/cuda/primitive/cast.cu [hunk 2]
```diff
@@ -31,7 +31,10 @@ struct CastFunctor {
 };
 
 template<typename To>
-struct CastFunctor<To, half, typename std::enable_if<!std::is_same<To, half>::value>::type> {
+struct CastFunctor<
+    To, half,
+    typename std::enable_if<!(std::is_same<To, half>::value || std::is_same<To, cuComplex>::value
+                              || std::is_same<To, cuDoubleComplex>::value)>::type> {
```
Comment (levi131) [inline oneflow/core/ep/cuda/primitive/cast.cu:37]: 有个小疑问，这里的`CastFunctor<To, half, typename std::enable_if<!std::is_same<To, half>::value>::type>` 和文件`unary_functor.cuh`中的`UnaryFunctor<DeviceType::kCUDA, UnaryOp::kCast, Dst, half>`有关系么。看起来在做同样的事情，后者比较完备，前者没有针对所有情况进行特化。

Comment (MarioLulab) [reply oneflow/core/ep/cuda/primitive/cast.cu:37]: 前者 CastFunctor 是 primitive 的一种，其可以在其他的 Kernel，如`ReduceSumHalfKernel`, `GpuFusedGruCellGradHalfKernel` 中被调用，这些 kernel 通常输入是 half 和 nv_bfloat16 的数据类型，需要通过 CastFunctor 来进行类型转换。后者 UnaryFunctor<DeviceType::kCUDA, UnaryOp::kCast, Dst, half> 是 "cast" 这个 op 最终会对应的调用的 functor，在 functioanl 层面使用 cast 这个函数，最终会调用的是这个UnaryFunctor<DeviceType::kCUDA, UnaryOp::kCast, Dst, Src>. 当前 pr 并没有调用到这里的 CastFunctor，会在后续的 commit 中删除该文件的更改

**Review by daquexian — APPROVED** [review]

File: python/oneflow/test/tensor/test_complex.py [hunk 3]
```diff
@@ -376,7 +399,36 @@ def test_add(self):
             compare_result(flow_x.grad.numpy(), np.ones(input_shape), 1e-5, 1e-2)
             compare_result(flow_y.grad.numpy(), np.ones(input_shape), 1e-5, 1e-2)
 
-    def test_sub(self):
+    @unittest.skipIf(os.getenv("ONEFLOW_TEST_CPU_ONLY"), "only test cpu cases")
+    def test_add_cuda(self):
```
[the remark this answers is not in the document]
Comment (MarioLulab) [reply python/oneflow/test/tensor/test_complex.py:403]: 修改 autotest 以支持 complex tensor 测试，将放在另一个 pr 里：
https://github.com/Oneflow-Inc/oneflow/pull/10027

**Review by levi131 — APPROVED** [review]
LGTM

**Pull request merged**